challenge_dishwasher/src/challenge_dishwasher/custom_find_cup.py

```diff
@@ -0,0 +1,213 @@
+import math
+
+import numpy as np
+import rospy
+import tf2_geometry_msgs
+from geometry_msgs.msg import PointStamped, Point
+from robot_skills.amigo import Amigo
+from sensor_msgs.msg import LaserScan
+from smach import StateMachine, State
+from tf import TransformListener
+from tf2_ros import Buffer
+from visualization_msgs.msg import MarkerArray, Marker
+
+_ = tf2_geometry_msgs
+
+
+class CustomFindCup(State):
+    def __init__(self, robot, box_offset_x=0.5, box_size_y=0.9, box_size_x=0.5):
+        State.__init__(self, outcomes=['succeeded', 'failed'], output_keys=['position'])
+        self._visualization_publisher = rospy.Publisher('find_cup_visualization', MarkerArray, queue_size=1)
+
+        self._robot = robot
+        self._box_offset_x = box_offset_x
+        self._box_size_y = box_size_y
+        self._box_size_x = box_size_x
+
+        self.tf_buffer = Buffer()
+        self.listener = TransformListener(self.tf_buffer)
+
+    @staticmethod
+    def _get_points_from_scan_msg(msg):
+        scan_points = []
+        for i, r in enumerate(msg.ranges):
+            if math.isnan(r):
+                continue
+
+            x = r * math.cos(msg.angle_min + i * msg.angle_increment)
+            y = r * math.sin(msg.angle_min + i * msg.angle_increment)
+            scan_points.append((x, y))
+        return scan_points
+
+    def _extract_cup_pose(self, exclude_points=[], exclude_radius=0.05, cluster_max_distance=0.05,
+                          cluster_min_points=10):
+        def distance(p1, p2):
+            return math.hypot(p1[0] - p2[0], p1[1] - p2[1])
+
+        def to_close_to_one_of_these_points(p, points, radius):
+            for e_p in points:
+                if distance(e_p, p) < radius:
+                    return True
+            return False
+
+        msg = rospy.wait_for_message("/amigo/torso_laser/scan", LaserScan)
```

Comment:
I know that this code is developed for AMIGO, but please make it as generic as possible for now.

Reply:
Better rewrite the whole thing then 😜